src/resolve/resolved-dns-cache.c

```diff
@@ -533,6 +546,10 @@ static int dns_cache_put_negative(
         assert(key);
         assert(owner_address);
 
+        /* Don't store the negative responses if the StaleRetentionSec is greater than zero. */
+        if (stale_retention_usec > 0 && rcode != DNS_RCODE_NXDOMAIN)
+                        return 0;
+
```

Comment:
No, don't do this. Below, the rcode is filtered down to either success, nxdomain, or servfail. The RFC explicitly states either success or nxdomain *must* replace our cached entry. So only when the rcode is servfail, and serve_stale is enabled, should we return 0. I think you should update the if check below.